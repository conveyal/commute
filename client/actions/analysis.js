import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  const commuters = newAnalysis.commuters
  delete newAnalysis.commuters
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`),
    makeMockTrips(newAnalysis.id, organizationId, commuters)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organizationId) => [
  deleteLocally({ analysisId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]

const makeMockTrips = (analysisId, organizationId, commuters) => {
  const trips = []
  const modes = ['bike', 'car', 'transit', 'walk']
  const randRange = () => Math.random() * 0.4 + 0.8
  commuters.forEach((commuter) => {
    const curTrip = {
      commuterId: commuter.id
    }
    const tripMetricBase = Math.random()
    let bestValue = 9999999
    let bestMode
    modes.forEach((mode) => {
      // generate random trips
      const modePctMultiplier = randRange()
      const cost = tripMetricBase * 34.7 * modePctMultiplier * randRange()
      const distance = tripMetricBase * 69 * modePctMultiplier * randRange()
      const time = tripMetricBase * 7500 * modePctMultiplier * randRange()
      curTrip[mode] = {cost, distance, time}
      const tripValue = cost + time * 0.01
      if (tripValue < bestValue) bestMode = mode
    })
    curTrip.mostLikely = Object.assign({ mode: bestMode }, curTrip[bestMode])
    trips.push(curTrip)
  })
  return sendMockTrips({analysisId, organizationId, trips})
}
const sendMockTrips = createAction('receive mock calculated trips')
