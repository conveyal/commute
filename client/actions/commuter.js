import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addCommuter = createAction('add commuter')
export const createCommuter = ({commuter, groupId, organizationId}) => {
  commuter.id = uuid.v4()
  return [
    addCommuter({commuter, groupId, organizationId}),
    push(`/organizations/${organizationId}/groups/${groupId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete commuter')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteCommuter = ({commuterId, groupId, organizationId}) => [
  deleteLocally({commuterId, groupId, organizationId}),
  push(`/organizations/${organizationId}/groups/${groupId}`)
]

const updateLocally = createAction('update commuter')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateCommuter = ({commuter, groupId, organizationId}) => [
  updateLocally({commuter, groupId, organizationId}),
  push(`/organizations/${organizationId}/groups/${groupId}`)
]
