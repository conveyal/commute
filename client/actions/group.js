import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addGroup = createAction('add group')
const appendCommuters = createAction('append commuters')
export const appendToGroup = (newCommuters, groupId, organizationId) => {
  return [
    appendCommuters({ commuters: newCommuters, groupId, organizationId }),
    push(`/organizations/${organizationId}/groups/${groupId}`)
  ] // TODO save to server
}
export const createGroup = (newGroup, organizationId) => {
  newGroup.id = uuid.v4()
  newGroup.allAddressesGeocoded = false
  return [
    addGroup({ group: newGroup, organizationId }),
    push(`/organizations/${organizationId}/groups/${newGroup.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete group')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/group/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteGroup = (id, organizationId) => [
  deleteLocally({ groupId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]
