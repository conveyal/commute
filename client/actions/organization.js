import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add organization')
export const createOrganization = (newOrganization) => {
  newOrganization.id = uuid.v4()
  newOrganization.analyses = []
  newOrganization.analysesById = {}
  newOrganization.sites = []
  newOrganization.sitesById = {}
  newOrganization.groups = []
  newOrganization.groupsById = {}
  return [
    addLocally(newOrganization),
    push(`/organizations/${newOrganization.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete organization')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/organization/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteOrganization = (id) => [deleteLocally(id), push('/')]

export const updateOrganization = (organization) => [
  setLocally(organization),
  push(`/organizations/${organization.id}`)
]
/* const updateOnServer = (organization) =>
  serverAction({
    url: `/api/organization/${organization.id}`,
    options: {
      body: JSON.stringify(organization),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set organization')
