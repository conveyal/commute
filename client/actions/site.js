import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addSite = createAction('add site')
export const createSite = (newSite, organizationId) => {
  newSite.id = uuid.v4()
  return [
    addSite({ organizationId, site: newSite }),
    push(`/organizations/${organizationId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete site')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/site/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteSite = (id, organizationId) => [
  deleteLocally({ siteId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]

const updateLocally = createAction('update site')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/site/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateSite = (site, organizationId) => [
  updateLocally({ organizationId, site }),
  push(`/organizations/${organizationId}`)
]
