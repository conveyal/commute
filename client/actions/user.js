import {createAction} from 'redux-actions'

// login / logout
export const login = createAction('log in')
export const logout = createAction('log out')
