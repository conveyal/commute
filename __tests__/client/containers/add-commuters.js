/* global describe, expect, File, it */

import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'

import {expectAppendCommuters, expectCreateGroup} from '../../test-utils/actions'
import {makeMockStore, mockStores} from '../../test-utils/mock-data.js'

import AddCommuters from '../../../client/containers/add-commuters'

const mockCsvFile = new File(['id,name,email,address\n1,Bob,a@b.c,"123 Main St"'], 'mockFile.csv')

describe('Container > AddCommuters', () => {
  it('Add Commuters View loads (create mode)', () => {
    const mockStore = makeMockStore(mockStores.oneSimpleOrganization)

    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <AddCommuters
          params={{organizationId: '1'}}
          />
      </Provider>
    )
    expect(mountToJson(tree.find(AddCommuters))).toMatchSnapshot()
  })

  it('Add Commuters View loads (with existing commuter group)', () => {
    // Given a logged-in user
    const mockStore = makeMockStore(mockStores.complexOrganization)

    // When the Add Commuters View is to be loaded
    // And the previous view was the Commuter Group View
    const tree = mount(
      <Provider store={mockStore}>
        <AddCommuters
          params={{organizationId: '2', groupId: '1'}}
          />
      </Provider>
    )

    // Then the Add Commuters View should load
    // And the commuter group name field should be populated and disabled
    // And there should be a button to go back to the Commuter Group
    expect(mountToJson(tree.find(AddCommuters))).toMatchSnapshot()
  })

  it('Preview Add Commuters', (done) => {
    // Given a logged-in user is viewing the Add Commuters View
    const mockStore = makeMockStore(mockStores.oneSimpleOrganization)
    const tree = mount(
      <Provider store={mockStore}>
        <AddCommuters
          params={{organizationId: '1'}}
          />
      </Provider>
    )

    // When a user selects a file to be uploaded
    // Then the file should be parsed
    tree.find('Dropzone').props().onDrop([mockCsvFile])

    // jsdom's filereader is asyncrhonous, so wait til it finishes
    setTimeout(() => {
      // And if the file is valid a preview of the commuters should
      // be shown in a table within an accordion
      expect(mountToJson(tree.find(AddCommuters))).toMatchSnapshot()
      done()
    }, 1000)
  })

  it('Create Commuter Group', () => {
    // Given a logged-in user is viewing the Add Commuters View
    const mockStore = makeMockStore(mockStores.oneSimpleOrganization)
    const tree = mount(
      <Provider store={mockStore}>
        <AddCommuters
          params={{organizationId: '1'}}
          />
      </Provider>
    )

    // When the user fills out all of the required fields (Name)
    tree.find('input').first().simulate('change', {target: {value: 'Mock Name'}})

    // And the user submits the form
    tree.find('form').find('button').simulate('click')

    expectCreateGroup(mockStore.getActions())
  })

  it('Append Commuters to existing commuter group', (done) => {
    // Given a logged-in user is viewing the Add Commuters View
    // And the user is adding commuters to an existing group
    const mockStore = makeMockStore(mockStores.complexOrganization)
    const tree = mount(
      <Provider store={mockStore}>
        <AddCommuters
          params={{organizationId: '2', groupId: '1'}}
          />
      </Provider>
    )

    // When a user selects a file to be uploaded
    tree.find('Dropzone').props().onDrop([mockCsvFile])

    // jsdom's filereader is asyncrhonous, so wait til it finishes
    setTimeout(() => {
      // And the user submits the form
      tree.find('form').find('button').simulate('click')

      expectAppendCommuters(mockStore.getActions())
      done()
    }, 1000)
  })
})
