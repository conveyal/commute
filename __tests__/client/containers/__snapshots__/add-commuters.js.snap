// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container > AddCommuters Add Commuters View loads 1`] = `
"<div class=\\"container\\">
  <div class=\\"row\\">
    <div class=\\"col-xs-12\\">
      <h3><span>Add Commuters to Site</span> 
        <button type=\\"button\\" class=\\"pull-right btn btn-default\\">
          <i class=\\"fa fa-arrow-left fa-fw \\"></i><span>Back</span> </button>
      </h3>
      <div class=\\"form-group\\">
        <label for=\\"site-label\\" class=\\"control-label\\">Site</label>
        <input type=\\"text\\" value=\\"Acme Corp\\" disabled=\\"\\" id=\\"site-label\\" class=\\"form-control\\">
      </div>
      <div class=\\"dropzone\\">
        <div>Try dragging and dropping a csv file here, or click to select files to upload. Make sure the csv file contains the headers: 'name' and 'address'.</div>
        <input type=\\"file\\" accept=\\"text/csv\\" style=\\"display: none;\\">
      </div>
      <div role=\\"tablist\\" class=\\"panel-group\\">
        <div class=\\"panel panel-info\\">
          <div class=\\"panel-heading\\">
            <h4 role=\\"presentation\\" class=\\"panel-title\\">
              <a role=\\"tab\\" aria-expanded=\\"false\\" aria-selected=\\"false\\" class=\\"collapsed\\">1 Existing Commuters</a> 
            </h4>
          </div>
          <div role=\\"tabpanel\\" class=\\"panel-collapse collapse\\" aria-hidden=\\"true\\">
            <div class=\\"panel-body\\">
              <div class=\\"react-bs-table-container\\">
                <div class=\\"react-bs-table\\" style=\\"height: 100%;\\">
                  <div class=\\"react-bs-container-header table-header-wrapper\\">
                    <table class=\\"table table-hover table-bordered\\">
                      <colgroup>
                        <col class=\\"\\" style=\\"display: none; min-width: NaNpx;\\">
                          <col class=\\"\\" style=\\"min-width: NaNpx;\\">
                            <col class=\\"\\" style=\\"min-width: NaNpx;\\">
                      </colgroup>
                      <thead>
                        <tr>
                          <th class=\\"\\" data-is-only-head=\\"false\\" style=\\"text-align: left; display: none;\\" data-field=\\"_id\\">
                            <div></div>
                          </th>
                          <th class=\\"\\" data-is-only-head=\\"false\\" title=\\"Name\\" style=\\"text-align: left;\\" data-field=\\"name\\">

                            <!-- react-text: 35 -->Name<!-- /react-text -->
                            <div></div>
                          </th>
                          <th class=\\"\\" data-is-only-head=\\"false\\" title=\\"Address\\" style=\\"text-align: left;\\" data-field=\\"address\\">

                            <!-- react-text: 38 -->Address<!-- /react-text -->
                            <div></div>
                          </th>
                        </tr>
                      </thead>
                    </table>
                  </div>
                  <div class=\\"react-bs-container-body\\" style=\\"height: 100%;\\">
                    <table class=\\"table table-bordered\\">
                      <colgroup>
                        <col class=\\"\\" style=\\"display: none;\\">
                          <col class=\\"\\">
                            <col class=\\"\\">
                      </colgroup>
                      <tbody>
                        <tr class=\\"\\">
                          <td class=\\"\\" style=\\"text-align: left; display: none;\\">commuter-2</td>
                          <td class=\\"\\" style=\\"text-align: left;\\">Luke Warm</td>
                          <td class=\\"\\" style=\\"text-align: left;\\">4321 XYZ Boulevard</td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Container > AddCommuters Append Commuters to existing site 1`] = `
Object {
  "payload": Object {
    "args": Array [
      "/site/site-2",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;

exports[`Container > AddCommuters Preview Add Commuters 1`] = `
"<div class=\\"container\\">
  <div class=\\"row\\">
    <div class=\\"col-xs-12\\">
      <h3><span>Add Commuters to Site</span> 
        <button type=\\"button\\" class=\\"pull-right btn btn-default\\">
          <i class=\\"fa fa-arrow-left fa-fw \\"></i><span>Back</span> </button>
      </h3>
      <div class=\\"form-group\\">
        <label for=\\"site-label\\" class=\\"control-label\\">Site</label>
        <input type=\\"text\\" value=\\"Acme Corp\\" disabled=\\"\\" id=\\"site-label\\" class=\\"form-control\\">
      </div>
      <div class=\\"dropzone\\">
        <div>Try dragging and dropping a csv file here, or click to select files to upload. Make sure the csv file contains the headers: 'name' and 'address'.</div>
        <input type=\\"file\\" accept=\\"text/csv\\" style=\\"display: none;\\">
      </div>
      <div role=\\"tablist\\" class=\\"panel-group\\">
        <div class=\\"panel panel-info\\">
          <div class=\\"panel-heading\\">
            <h4 role=\\"presentation\\" class=\\"panel-title\\">
              <a role=\\"tab\\" aria-expanded=\\"false\\" aria-selected=\\"false\\" class=\\"collapsed\\">1 Existing Commuters</a> 
            </h4>
          </div>
          <div role=\\"tabpanel\\" class=\\"panel-collapse collapse\\" aria-hidden=\\"true\\">
            <div class=\\"panel-body\\">
              <div class=\\"react-bs-table-container\\">
                <div class=\\"react-bs-table\\" style=\\"height: 100%;\\">
                  <div class=\\"react-bs-container-header table-header-wrapper\\">
                    <table class=\\"table table-hover table-bordered\\">
                      <colgroup>
                        <col class=\\"\\" style=\\"display: none; min-width: NaNpx;\\">
                          <col class=\\"\\" style=\\"min-width: NaNpx;\\">
                            <col class=\\"\\" style=\\"min-width: NaNpx;\\">
                      </colgroup>
                      <thead>
                        <tr>
                          <th class=\\"\\" data-is-only-head=\\"false\\" style=\\"text-align: left; display: none;\\" data-field=\\"_id\\">
                            <div></div>
                          </th>
                          <th class=\\"\\" data-is-only-head=\\"false\\" title=\\"Name\\" style=\\"text-align: left;\\" data-field=\\"name\\">

                            <!-- react-text: 35 -->Name<!-- /react-text -->
                            <div></div>
                          </th>
                          <th class=\\"\\" data-is-only-head=\\"false\\" title=\\"Address\\" style=\\"text-align: left;\\" data-field=\\"address\\">

                            <!-- react-text: 38 -->Address<!-- /react-text -->
                            <div></div>
                          </th>
                        </tr>
                      </thead>
                    </table>
                  </div>
                  <div class=\\"react-bs-container-body\\" style=\\"height: 100%;\\">
                    <table class=\\"table table-bordered\\">
                      <colgroup>
                        <col class=\\"\\" style=\\"display: none;\\">
                          <col class=\\"\\">
                            <col class=\\"\\">
                      </colgroup>
                      <tbody>
                        <tr class=\\"\\">
                          <td class=\\"\\" style=\\"text-align: left; display: none;\\">commuter-2</td>
                          <td class=\\"\\" style=\\"text-align: left;\\">Luke Warm</td>
                          <td class=\\"\\" style=\\"text-align: left;\\">4321 XYZ Boulevard</td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class=\\"panel panel-success\\">
          <div class=\\"panel-heading\\">
            <h4 role=\\"presentation\\" class=\\"panel-title\\">
              <a role=\\"tab\\" aria-expanded=\\"false\\" aria-selected=\\"false\\" class=\\"collapsed\\">1 New Commuters</a> 
            </h4>
          </div>
          <div role=\\"tabpanel\\" class=\\"panel-collapse collapse\\" aria-hidden=\\"true\\">
            <div class=\\"panel-body\\">
              <div class=\\"react-bs-table-container\\">
                <div class=\\"react-bs-table\\" style=\\"height: 100%;\\">
                  <div class=\\"react-bs-container-header table-header-wrapper\\">
                    <table class=\\"table table-hover table-bordered\\">
                      <colgroup>
                        <col class=\\"\\" style=\\"display: none; min-width: NaNpx;\\">
                          <col class=\\"\\" style=\\"min-width: NaNpx;\\">
                            <col class=\\"\\" style=\\"min-width: NaNpx;\\">
                      </colgroup>
                      <thead>
                        <tr>
                          <th class=\\"\\" data-is-only-head=\\"false\\" style=\\"text-align: left; display: none;\\" data-field=\\"_id\\">
                            <div></div>
                          </th>
                          <th class=\\"\\" data-is-only-head=\\"false\\" title=\\"Name\\" style=\\"text-align: left;\\" data-field=\\"name\\">

                            <!-- react-text: 70 -->Name<!-- /react-text -->
                            <div></div>
                          </th>
                          <th class=\\"\\" data-is-only-head=\\"false\\" title=\\"Address\\" style=\\"text-align: left;\\" data-field=\\"address\\">

                            <!-- react-text: 73 -->Address<!-- /react-text -->
                            <div></div>
                          </th>
                        </tr>
                      </thead>
                    </table>
                  </div>
                  <div class=\\"react-bs-container-body\\" style=\\"height: 100%;\\">
                    <table class=\\"table table-bordered\\">
                      <colgroup>
                        <col class=\\"\\" style=\\"display: none;\\">
                          <col class=\\"\\">
                            <col class=\\"\\">
                      </colgroup>
                      <tbody>
                        <tr class=\\"\\">
                          <td class=\\"\\" style=\\"text-align: left; display: none;\\">1</td>
                          <td class=\\"\\" style=\\"text-align: left;\\">Bob</td>
                          <td class=\\"\\" style=\\"text-align: left;\\">123 Main St</td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <button type=\\"button\\" class=\\"btn btn-success\\">Add Commuters</button>
    </div>
  </div>
</div>"
`;
